File: java/testsrc/tako/TestConversions.java
// Copyright 2020 Jacob Glueck
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package tako;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import tako.ParseException;
import takogen.test_types.Conversions.FlavorOld;
import takogen.test_types.Conversions.FlavorNew;
import takogen.test_types.Conversions.CupcakeOrderOld;
import takogen.test_types.Conversions.CupcakeOrderNew;
import takogen.test_types.Conversions.CakeOrderOld;
import takogen.test_types.Conversions.CakeOrderNew;
import takogen.test_types.Conversions.CakeOrderOldView;
import takogen.test_types.Conversions.CakeOrderNewView;
import takogen.test_types.Conversions.OrderOld;
import takogen.test_types.Conversions.OrderNew;
import takogen.test_types.Conversions.MsgOld;
import takogen.test_types.Conversions.MsgNew;
import static takogen.test_types.Conversions.convert;
import static tako.Helpers.*;

public class TestConversions {
    @Test
    public void flavorOldToFlavorNew() {
        var old = FlavorOld.CHOCOLATE;
        var new_ = FlavorNew.makeUninitialized();
        convert(new_, old);
        Assertions.assertEquals(new_, FlavorNew.CHOCOLATE);
    }
    @Test
    public void flavorNewToFlavorOld() {
        var new_ = FlavorNew.CARMEL;
        var old = FlavorOld.makeUninitialized();
        convert(old, new_);
        Assertions.assertEquals(old, FlavorOld.CHOCOLATE);
    }
    @Test
    public void cupcakeOrderOldToCupcakeOrderNew() {
        var old = new CupcakeOrderOld().init()
            .flavor(FlavorOld.CHOCOLATE)
            .finish();
        var new_ = new CupcakeOrderNew();
        convert(new_, old);
        var expected = new CupcakeOrderNew().init()
            .flavor(FlavorNew.CHOCOLATE)
            .quantity(50)
            .finish();
        Assertions.assertEquals(new_, expected);
    }
    @Test
    public void cupcakeOrderNewToCupcakeOrderOld() {
        var new_ = new CupcakeOrderNew().init()
            .flavor(FlavorNew.CARMEL)
            .quantity(50)
            .finish();
        var old = new CupcakeOrderOld();
        convert(old, new_);
        var expected = new CupcakeOrderOld().init()
            .flavor(FlavorOld.CHOCOLATE)
            .finish();
        Assertions.assertEquals(old, expected);
    }
    @Test
    public void orderOldToOrderNew() {
        var old = new OrderOld();
        old.set(CupcakeOrderOld.class).init()
            .flavor(FlavorOld.CHOCOLATE)
            .finish();
        var new_ = new OrderNew();
        convert(new_, old);
        var expected = new OrderNew();
        expected.set(CupcakeOrderNew.class).init()
            .flavor(FlavorNew.CHOCOLATE)
            .quantity(50)
            .finish();
        Assertions.assertEquals(new_, expected);
    }
    @Test
    public void orderNewToOrderOld() {
        var new_ = new OrderNew();
        new_.set(CupcakeOrderNew.class).init()
            .flavor(FlavorNew.CARMEL)
            .quantity(50)
            .finish();
        var old = new OrderOld();
        convert(old, new_);
        var expected = new OrderOld();
        expected.set(CupcakeOrderOld.class).init()
            .flavor(FlavorOld.CHOCOLATE)
            .finish();
        Assertions.assertEquals(old, expected);
    }
    @Test
    public void msgOldToMsgNew() {
        var old = new MsgOld();
        old.msg().set(CupcakeOrderOld.class).init()
            .flavor(FlavorOld.CHOCOLATE);
        var new_ = new MsgNew();
        convert(new_, old);
        var expected = new MsgNew();
        expected.msg().set(CupcakeOrderNew.class).init()
            .flavor(FlavorNew.CHOCOLATE)
            .quantity(50);
        Assertions.assertEquals(new_, expected);
    }
    @Test
    public void msgNewToMsgOld() {
        var new_ = new MsgNew();
        new_.msg().set(CupcakeOrderNew.class).init()
            .flavor(FlavorNew.CARMEL)
            .quantity(50);
        var old = new MsgOld();
        convert(old, new_);
        var expected = new MsgOld();
        expected.msg().set(CupcakeOrderOld.class).init()
            .flavor(FlavorOld.CHOCOLATE);
        Assertions.assertEquals(old, expected);
    }
    @Test
    public void viewCakeOrderOldToCakeOrderNew() throws ParseException {
        var old = new CakeOrderOld().init()
            .flavor(FlavorOld.CHOCOLATE)
            .finish();
        var oldBytes = old.serialize();
        var oldView = new CakeOrderOldView();
        oldView.parse(oldBytes, 0);
        var newView = new CakeOrderNewView();
        convert(newView, oldView);
        var new_ = new CakeOrderNew().init()
            .flavor(FlavorNew.CHOCOLATE)
            .finish();
        var newViewBuilt = new CakeOrderNew();
        newView.build(newViewBuilt);
        Assertions.assertEquals(newViewBuilt, new_);
    }
}

File: pytest/test_conversions.py
# Copyright 2020 Jacob Glueck
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from unittest import TestCase

from takogen.test_types.Conversions import *


class TestConversions(TestCase):
    def test_conversions_flavor_old_to_flavor_new(self) -> None:
        old = FlavorOld.CHOCOLATE
        new_ = convert(old, ToFlavorNew)
        self.assertEqual(new_, FlavorNew.CHOCOLATE)

    def test_conversions_flavor_new_to_flavor_old(self) -> None:
        new_ = FlavorNew.CARMEL
        old = convert(new_, ToFlavorOld)
        self.assertEqual(old, FlavorOld.CHOCOLATE)

    def test_conversions_cupcake_order_old_to_cupcacke_order_new(self) -> None:
        old = CupcakeOrderOld(flavor=FlavorOld.CHOCOLATE)
        new_ = convert(old, ToCupcakeOrderNew)
        self.assertEqual(new_, CupcakeOrderNew(flavor=FlavorNew.CHOCOLATE, quantity=50))

    def test_conversions_cupcake_order_new_to_cupcacke_order_old(self) -> None:
        new_ = CupcakeOrderNew(flavor=FlavorNew.CARMEL, quantity=50)
        old = convert(new_, ToCupcakeOrderOld)
        self.assertEqual(old, CupcakeOrderOld(flavor=FlavorOld.CHOCOLATE))

    def test_conversions_order_old_to_order_new(self) -> None:
        old = OrderOld(CupcakeOrderOld(flavor=FlavorOld.CHOCOLATE))
        new_ = convert(old, ToOrderNew)
        self.assertEqual(
            new_, OrderNew(CupcakeOrderNew(flavor=FlavorNew.CHOCOLATE, quantity=50))
        )

    def test_conversions_order_new_to_order_old(self) -> None:
        new_ = OrderNew(CupcakeOrderNew(flavor=FlavorNew.CARMEL, quantity=50))
        old = convert(new_, ToOrderOld)
        self.assertEqual(old, OrderOld(CupcakeOrderOld(flavor=FlavorOld.CHOCOLATE)))

    def test_conversions_msg_old_to_msg_new(self) -> None:
        old = MsgOld(OrderOld(CupcakeOrderOld(flavor=FlavorOld.CHOCOLATE)))
        new_ = convert(old, ToMsgNew)
        self.assertEqual(
            new_,
            MsgNew(OrderNew(CupcakeOrderNew(flavor=FlavorNew.CHOCOLATE, quantity=50))),
        )

    def test_conversions_msg_new_to_msg_old(self) -> None:
        new_ = MsgNew(OrderNew(CupcakeOrderNew(flavor=FlavorNew.CARMEL, quantity=50)))
        old = convert(new_, ToMsgOld)
        self.assertEqual(
            old, MsgOld(OrderOld(CupcakeOrderOld(flavor=FlavorOld.CHOCOLATE)))
        )
